flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/channel/ChannelStateWriterImpl.java

```diff
@@ -0,0 +1,225 @@
+		rethrow();
+		Preconditions.checkState(isRunning && asyncWriter.isAlive(), "not running");
+		ChannelStateWriteResult result = new ChannelStateWriteResult();
+		ChannelStateWriteResult put = results.computeIfAbsent(checkpointId, id -> {
+			Preconditions.checkArgument(results.size() < maxCheckpoints, "results.size() > maxCheckpoints", results.size(), maxCheckpoints);
+			enqueue(new CheckpointStartRequest(checkpointId, result, checkpointOptions.getTargetLocation()));
+			return result;
+		});
+		Preconditions.checkArgument(put == result, "result future already present for checkpoint id: " + checkpointId);
+	}
+
+	@Override
+	public void addInputData(long checkpointId, InputChannelInfo info, int startSeqNum, Buffer... data) {
+		LOG.debug("add input data, checkpoint id: {}, channel: {}, startSeqNum: {}, num buffers: {}",
+			checkpointId, info, startSeqNum, data == null ? 0 : data.length);
+		enqueue(write(checkpointId, info, checkBufferType(data)));
+	}
+
+	@Override
+	public void addOutputData(long checkpointId, ResultSubpartitionInfo info, int startSeqNum, Buffer... data) {
+		LOG.debug("add output data, checkpoint id: {}, channel: {}, startSeqNum: {}, num buffers: {}",
+			checkpointId, info, startSeqNum, data == null ? 0 : data.length);
+		enqueue(write(checkpointId, info, checkBufferType(data)));
+	}
+
+	@Override
+	public void finishInput(long checkpointId) {
+		LOG.debug("finish input data, checkpoint id: {}", checkpointId);
+		enqueue(completeInput(checkpointId));
+	}
+
+	@Override
+	public void finishOutput(long checkpointId) {
+		LOG.debug("finish output data, checkpoint id: {}", checkpointId);
+		enqueue(completeOutput(checkpointId));
+	}
+
+	@Override
+	public ChannelStateWriteResult getWriteResult(long checkpointId) {
+		LOG.debug("requested write result, checkpoint id: {}", checkpointId);
+		ChannelStateWriteResult result = results.remove(checkpointId);
+		Preconditions.checkArgument(result != null, "channel state write result not found for checkpoint id " + checkpointId);
+		return result;
+	}
+
+	public void open() {
+		Preconditions.checkState(isRunning);
+		asyncWriter.start();
+	}
+
+	@Override
+	public void close() {
+		isRunning = false;
+		results.clear();
+		asyncWriter.interrupt();
+		while (asyncWriter.isAlive()) {
+			try {
+				asyncWriter.join();
+			} catch (InterruptedException e) {
+				if (!asyncWriter.isAlive()) {
+					Thread.currentThread().interrupt();
+				}
+				LOG.debug("interrupted while waiting for the writer thread to die", e);
+			}
+		}
+		rethrow();
+	}
+
+	private void enqueue(ChannelStateWriteRequest request) {
+		rethrow();
+		Preconditions.checkState(isRunning);
+		try {
+			handover.put(request);
+		} catch (InterruptedException e) {
+			throw new RuntimeException("Interrupted while trying to add new handover request", e);
+		}
+	}
+
+	private void rethrow() {
+		if (thrown != null) {
+			throw new RuntimeException(thrown); // wrap to record current stack-trace
```

Comment:
I do not think `RuntimeException` is a good way unless necessary. Although it seems simple to not announce exceptions explicitly in related methods, it might mislead the upper caller to think all the interactions with `ChannelStateWriter` would not cause any exceptions. Then the caller might lose the chance to handle exceptions in elegant way.

Also the `RuntimeException` seems have a more widely concept, I guess it probably should be `IOException` for writer to give more sense.

Reply:
I don't see how the client could recover from any exception thrown by this class. Especially given that: 
1. these exceptions can come from previous calls
1. the idiomatic way to handle errors in Flink (as I see it) is to fail and recover the whole job

Also, it can be any exception, not only `IOException`.

So there is no point to declare (likely forcing the client to wrap it or declare too).

Reply:
Yes, I agree that any exceptions should fail the task and result in restarting the whole job.

My previous concern was that it may be better to announce the specific exceptions explicitly for the called methods, to make the caller have the possibility to handle the exception in elegant way.

E.g. If the internal called methods without explicit exception:
#method1() {
     action 1: request some resources  
     action 2: call some methods without exception
     action 3: release resources
}

E.g. If the internal called methods with explicit exception
#method1() {
    action 1: request some resources  
    try{
          action 2: call some methods without exception
    } finally {
          action 3: release resources
    }
}

For the first example, it might cause resource leak for the caller if is not ware of the exception.

I guess all the related exceptions for these methods should be `IOException` and `InterruptedException`.

Reply:
IMO, resources should be freed in `finally` regardless of exceptions declared in method signature.
Currently (#11507 PR), any exceptions are propagated, caught somewhere (as `Exception`) just to be wrapped and rethrown again. 
In one case there is `finally` block freeing resources; but it must free them regardless of this call.
So, there will be no changes in the caller code if we use checked exceptions (except for some problems with lambdas).
But I see checked exceptions are used in many places in Flink so I'm also fine with this option.

Reply:
Yes, I understand and agree with you points to some extent. It is indeed a bit trouble to carry the specific exception in many different components, and also for the case to wrapper it in most lambda cases.

My previous example was for forcing the caller to release resources within `try...finally` to pass compiling if we announce the exception explicitly in interface method. Otherwise some callers might not have the good habit to always release resources within finally region, especially in tests.

Another minor concern was that we only announce the `IllegalArgumentException` for interface methods, and actually hide the other `IOException` , `InteruphttedException` to wrapper implicitly. My guessing is that we think the `IllegalArgumentException` is more important to bring alerts for callers. :)